calcom/cal.com
# perf: improve typescript perf by pre-generating the router types

## What does this PR do?



### UPDATE

We ended up adding a turborepo task to run the build script on the trpc package.
This allows devs to run `yarn build` at the root of the project and have the types work fast. The web's build task now has a dependency on the trpc's package build script.

For a future task we should probably add support for typescript's project references to help typescript with the load even further.

### Notes

The typescript language server gets very slow / crashes often when working on the web.
I've tried to debug its perf by following these docs:
- https://github.com/microsoft/TypeScript/wiki/Performance#using-type-annotations
- https://github.com/microsoft/TypeScript/wiki/Debugging-Language-Service-in-VS-Code
- https://github.com/trpc/trpc/discussions/2448

Following microsoft's docs revealed the slowness to be somewhere around the trpc router in combination with the generated prisma types.

The command I ran was: 

```sh
cd apps/web
NODE_OPTIONS=--max_old_space_size=8096 yarn tsc -p . --generateTrace . --incremental false
```

![image](https://github.com/user-attachments/assets/504a90af-b4de-4cda-9bbf-6b7f14fde3e0)


It was hard to pinpoint exactly where the issue is but one of the major perf hits seems to be around the `useHasPaidPlan` hook which uses the `trpc.viewer.teams.hasTeamPlan` query which uses the `prisma.membership` object. I haven't worked with prisma or trpc before but the stack probably does a lot of heavy lifting with typescript to provide type-safety and autocompletion. It could be too much for tsserver to do on the fly.

I didn't explore all the possible solutions but the last link in the trpc repo had a solution that seems to improve the perf significantly, at least when working on the web (working on the trpc package is still slow).

- Fixes #15175
- Fixes CAL-3796



#### Notes
- It requires to generate the router types for every change to any route.
  - This is done with turborepo.

## Mandatory Tasks (DO NOT REMOVE)


## How should this be tested?

- Run `yarn build` from `packages/trpc/package.json` to generate the `types/router.d.ts` file.
- Open up a file in the web (e.g. `modules/event-types/views/event-types-listing-view.tsx`)
- Notice server boots up fast and types are inferred correctly, e.g. https://github.com/dlvhdr/cal.com/blob/main/apps/web/modules/event-types/views/event-types-listing-view.tsx?plain=1#L232

## Checklist
